x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/kerberos/support/KerberosTicketValidator.java

```diff
@@ -0,0 +1,254 @@
+
+            String base64OutToken = null;
+            if (outToken != null && outToken.length > 0) {
+                base64OutToken = Base64.getEncoder().encodeToString(outToken);
+            }
+            LOGGER.trace("validateTicket isGSSContextEstablished = {}, username = {}, outToken = {}", gssContext.isEstablished(),
+                    gssContext.getSrcName().toString(), base64OutToken);
+            return new Tuple<>(gssContext.isEstablished() ? gssContext.getSrcName().toString() : null, base64OutToken);
+        } catch (PrivilegedActionException pve) {
+            if (pve.getException() instanceof LoginException) {
+                throw (LoginException) pve.getCause();
+            }
+            if (pve.getException() instanceof GSSException) {
+                throw (GSSException) pve.getCause();
+            }
+            throw ExceptionsHelper.convertToRuntime((Exception) ExceptionsHelper.unwrapCause(pve));
+        } finally {
+            privilegedDisposeNoThrow(loginContext);
+            privilegedCloseNoThrow(gssContext);
+        }
+    }
+
+    private static byte[] acceptSecContext(final byte[] base64Ticket, GSSContext gssContext, LoginContext loginContext)
+            throws PrivilegedActionException {
+        final GSSContext finalGSSContext = gssContext;
+        // process token with gss context
+        return doAsWrapper(loginContext.getSubject(),
+                (PrivilegedExceptionAction<byte[]>) () -> finalGSSContext.acceptSecContext(base64Ticket, 0, base64Ticket.length));
+    }
+
+    private static GSSCredential createCredentials(final String servicePrincipalName, Oid nameType, final GSSManager gssManager,
+            LoginContext loginContext) throws GSSException, PrivilegedActionException {
+        final GSSName gssServicePrincipalName;
+        if (servicePrincipalName.equals("*") == false) {
+            gssServicePrincipalName = gssManager.createName(servicePrincipalName, nameType);
+        } else {
+            gssServicePrincipalName = null;
+        }
+        return doAsWrapper(loginContext.getSubject(), (PrivilegedExceptionAction<GSSCredential>) () -> gssManager
+                .createCredential(gssServicePrincipalName, GSSCredential.DEFAULT_LIFETIME, SPNEGO_OID, GSSCredential.ACCEPT_ONLY));
+    }
+
+    /**
+     * Privileged Wrapper that invokes action with Subject.doAs
+     *
+     * @param subject {@link Subject}
+     * @param action {@link PrivilegedExceptionAction} action for performing inside
+     *            Subject.doAs
+     * @return
+     * @throws PrivilegedActionException
+     */
+    private static <T> T doAsWrapper(final Subject subject, final PrivilegedExceptionAction<T> action) throws PrivilegedActionException {
+        try {
+            return AccessController.doPrivileged((PrivilegedExceptionAction<T>) () -> Subject.doAs(subject, action));
+        } catch (PrivilegedActionException pae) {
+            throw (PrivilegedActionException) pae.getException();
```

Comment:
do we need to unwrap here? Also, we should use `getCause` instead of `getException`. From the javadoc:

```java
    /**
     * Returns the exception thrown by the privileged computation that
     * resulted in this {@code PrivilegedActionException}.
     *
     * <p>This method predates the general-purpose exception chaining facility.
     * The {@link Throwable#getCause()} method is now the preferred means of
     * obtaining this information.
     *
     * @return the exception thrown by the privileged computation that
     *         resulted in this {@code PrivilegedActionException}.
     * @see PrivilegedExceptionAction
     * @see AccessController#doPrivileged(PrivilegedExceptionAction)
     * @see AccessController#doPrivileged(PrivilegedExceptionAction,
     *                                            AccessControlContext)
     */
```